%3a14/char-sets/inversion-list-check.scm

```diff
@@ -0,0 +1,151 @@
+; Part of Scheme 48 1.9.  See file COPYING for notices and license.
```

Comment:
Can you give me some background on this library and why we need it here instead of using something else? I am not sure how comfortable I feel embedding a fairly sophisticated library (or so it seems) directly into another library.

Reply:
Sure. The original SRFI-14 implementation only needs to deal with latin-1, which is very simple to do with 256 entry lookup tables. But with Unicode the simple lookup table approach is no longer feasible; the tables as just too large. Instead a data structure called an inversion list is used. This is a very common data structure for Unicode libraries which was popularized among implementers through the book "Unicode Demystified" (Gillam, 2002).

Inversion lists work by storing the ranges in a sorted array, the first entry being the lower value of the range and the second entry being one above the upper value of the range, and so on. The character set [A-Za-z] would be written as `#(#\A #\[ #\a #\{)`. Lookup is done through binary search; if the search terminates on an even index then the character is in the set. Set operations are quite simple, but I will not describe them.

Anyhow, the inversion list library exactly matches the use case for SRFI-14 and Scheme 48 probably doesn't use it for anything else. I used inversion lists myself when I added Unicode support in my own Scheme implementation. So should a bug appear, I can certainly fix it.